File: src/main/java/br/com/blu/bs2/sintetizador/contas/csv/model/dominio/Conta.java
package br.com.blu.bs2.sintetizador.contas.csv.model.dominio;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Data;
import lombok.Setter;

import java.math.BigDecimal;
import java.util.Objects;

@Data
@Builder
public class Conta {
    private Long numero;
    private Integer digito;
    private TipoConta tipoConta;
    @Setter(value = AccessLevel.PRIVATE)
    private BigDecimal saldo;

    public void addSaldo(BigDecimal saldo) {
        if (Objects.isNull(this.saldo))
            this.saldo = BigDecimal.ZERO;
        this.saldo = this.saldo.add(saldo);
    }

    public Conta adicionarSaldo(BigDecimal saldo) {
        addSaldo(saldo);
        return this;
    }

    public void gerarTipoConta() {
        if (Objects.isNull(this.digito) || this.digito < 0)
            return;
        this.tipoConta = TipoConta.of(this.digito);
    }

    public Conta geraTipoConta() {
        gerarTipoConta();
        return this;
    }

    public boolean isTipoContaValido() {
        return (Objects.nonNull(this.tipoConta));
    }

    public boolean isTipoContaIsComum() {
        return (isTipoContaValido() && this.tipoConta.isContaComum(this.digito));
    }

    public boolean isTipoContaIsPremium() {
        return (isTipoContaValido() && this.tipoConta.isContaPremium(this.digito));
    }

    public boolean isTipoContaIsExecutiva() {
        return (isTipoContaValido() && this.tipoConta.isContaExecutiva(this.digito));
    }

    public String toStringNumeroContaCompleta() {
        return (isParamGenerateNumeroContaCompleta())
                ? String.valueOf(this.numero).concat("-").concat(String.valueOf(this.digito))
                : "";
    }

    private boolean isParamGenerateNumeroContaCompleta() {
        return (isNumeroValido() && isDigitoContaValido());
    }

    private boolean isNumeroValido() {
        return Objects.nonNull(this.numero);
    }

    private boolean isDigitoContaValido() {
        return Objects.nonNull(this.digito);
    }
}

File: src/main/kotlin/br/com/blu/bs2/sintetizador/contas/csv/model/dominio/Conta.kt
package br.com.blu.bs2.sintetizador.contas.csv.model.dominio

import java.math.BigDecimal
import java.util.*

class Conta {
    var numero: Long? = null
    var digito: Int? = null
    var tipoConta: TipoConta? = null
    private var saldo: BigDecimal? = null
    fun getSaldo(): BigDecimal? {
        return saldo
    }

    fun setSaldo(saldo: BigDecimal?) {
        addSaldo(saldo)
    }

    fun addSaldo(saldo: BigDecimal?) {
        if (Objects.isNull(this.saldo)) this.saldo = BigDecimal.ZERO
        this.saldo = this.saldo!!.add(saldo)
    }

    fun adicionarSaldo(saldo: BigDecimal?): Conta {
        addSaldo(saldo)
        return this
    }

    fun gerarTipoConta() {
        if (Objects.isNull(digito) || digito!! < 0) return
        tipoConta = TipoConta.of(digito!!)
    }

    fun geraTipoConta(): Conta {
        gerarTipoConta()
        return this
    }

    val isTipoContaValido: Boolean
        get() = Objects.nonNull(tipoConta)
    val isTipoContaIsComum: Boolean
        get() = isTipoContaValido && tipoConta!!.isContaComum(digito!!)
    val isTipoContaIsPremium: Boolean
        get() = isTipoContaValido && tipoConta!!.isContaPremium(digito!!)
    val isTipoContaIsExecutiva: Boolean
        get() = isTipoContaValido && tipoConta!!.isContaExecutiva(digito!!)

    fun toStringNumeroContaCompleta(): String {
        return if (isParamGenerateNumeroContaCompleta) numero.toString() + "-" + digito.toString() else ""
    }

    private val isParamGenerateNumeroContaCompleta: Boolean
        get() = isNumeroValido && isDigitoContaValido
    private val isNumeroValido: Boolean
        get() = Objects.nonNull(numero)
    private val isDigitoContaValido: Boolean
        get() = Objects.nonNull(digito)

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is Conta) return false
        return numero == o.numero && digito == o.digito
    }

    override fun hashCode(): Int {
        return Objects.hash(numero, digito)
    }

    override fun toString(): String {
        return "Conta{" +
                "numero=" + numero +
                ", digito=" + digito +
                ", tipoConta=" + tipoConta +
                ", saldo=" + saldo +
                '}'
    }
}
